### PR TITLE
Session interface

### PR DESCRIPTION
**Pull request by puzza007**
WIP

@zoldar any thoughts on this?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -48,6 +48,23 @@ Req = #{url => <<"https://example.com">>.
...
+                               method => post,
+                               headers => ReqHeaders,
+                               connecttimeout_ms => 5000,
+                               proxy => <<"http://127.0.0.1:9000">>,
+                               sslverifyhost => false,
+                               sslverifypeer => false}.
+{{ok, #{status := 200}}, Session2} =
+    katipo_session:req(#{body => <<"some data">>}, Session).
+{{ok, #{status := 200}}, Session3} =
+    katipo_session:req(#{body => <<"different payload data">>}, Session3).
```
Comment (zoldar) [inline README.md]: Shouldn't it be `Session2`? `Session3` is unbound yet anyway.

File: src/katipo.erl [hunk 2]
```diff
@@ -558,7 +560,7 @@ opt(proxy, Proxy, {Req, Errors}) when is_binary(Proxy) ->
 opt(K, V, {Req, Errors}) ->
     {Req, [{K, V} | Errors]}.
 
--spec process_opts(map()) -> {ok, #req{}} | {error, {bad_opts, [any()]}}. %%todo
+-spec process_opts(request()) -> {ok, #req{}} | {error, {bad_opts, [any()]}}. %%todo
 process_opts(Opts) ->
     case maps:fold(fun opt/3, {#req{}, []}, Opts) of
         {Req=#req{}, []} ->
```
Comment (zoldar) [inline src/katipo.erl]: IMHO this should be a separate, refactoring related commit.

**Comment by puzza007** [comment]
Ta! :smile:

**Comment by zoldar** [comment]
Just a thought, but wouldn't it be beneficial to default to `hash` pooling when fetching workers in that mode? I'm not fully aware of implications right now, but persistent connections are one of the "selling points" of session-like interface. Correct me if I'm wrong of this but default `round_robin` interface causes establishing new connection for every request?

**Comment by puzza007** [comment]
I think you could still rely on the underlying application config (`hash`) if you wanted to maximise pipeline reuse. Perhaps an optimisation is if `hash` is set, you can assign the worker in `katipo_session:new` (there would be a little bit of fiddling to update it on worker death).

I'd definitely like to look at a way of setting pipelining per session, or something like that.

**Comment by zoldar** [comment]
I think that explicitly passing worker to `katipo_session:new` doesn't give much benefit - using `hash` pooling is enough - user will benefit from that as long as he makes calls to the same host/port combination.

**Comment by puzza007** [comment]
Ah I was thinking of doing it implicitly inside `:new`. Guess I should actually measure first :)

**Comment by zoldar** [comment]
Ah, right. True, true, better to measure than to speculate. Anyway, as it stands, :+1: from me ;)

File: test/katipo_SUITE.erl [hunk 3]
```diff
@@ -504,6 +522,75 @@ proxy_post_data(_) ->
...
+             headers => [{<<"header1">>, <<"!@#$%^&*()">>}]},
+    {ok, Session2} = katipo_session:update(Req2, Session),
+    {{ok, #{status := 200, body := Body}}, _Session3} =
+        katipo_session:req(#{}, Session2),
+    Json = jsx:decode(Body),
+    Expected =  [{<<"Accept">>,<<"*/*">>},
+                 {<<"Accept-Encoding">>,<<"gzip,deflate">>},
+                 {<<"Header1">>,<<"!@#$%^&*()">>},
+                 {<<"Host">>,<<"httpbin.org">>}],
+    [] = Expected -- proplists:get_value(<<"headers">>, Json).
```
Comment (ibnfirnas) [inline test/katipo_SUITE.erl]: 🏆

**Pull request merged**